File: python/queen.py
"""Solution to the Queen problem using Python."""

import argparse
from enum import Enum
from typing import List


class OccupancyStatus(Enum):
    """Used to represent the occupancy status of a column or diagonal."""

    FREE = 0
    OCCUPIED = 1


class QueenProblemSolver:
    """Solves the Queen problem for N Queens on NxN chessboard."""

    def __init__(self, n: int):
        self.n: int = n
        self._placed_queen_id: List[int] = [0 for _ in range(self.n)]
        self.solution_counter: int = 0
        self._init_occupancy()

        self._place_nth_queen(0)

    def _init_occupancy(self) -> None:
        self._is_occupied_column: List[OccupancyStatus] = [
            OccupancyStatus.FREE
        ] * self.n
        number_of_diagonals: int = self._get_number_of_diagonals()
        self._is_occupied_diagonal_right: List[OccupancyStatus] = [
            OccupancyStatus.FREE
        ] * number_of_diagonals
        self._is_occupied_diagonal_left: List[OccupancyStatus] = [
            OccupancyStatus.FREE
        ] * number_of_diagonals

    def _place_nth_queen(self, row: int) -> None:
        for column in range(self.n):
            if not self._is_occupied(row, column):
                self._placed_queen_id[row] = column
                self._update_occupancy(row, column, OccupancyStatus.OCCUPIED)

                if (row + 1) < self.n:
                    self._place_nth_queen(row + 1)
                else:
                    self._print_solution()

                self._update_occupancy(row, column, OccupancyStatus.FREE)

    def _is_occupied(self, row, column) -> bool:
        return OccupancyStatus.OCCUPIED in (
            self._is_occupied_column[column],
            self._is_occupied_diagonal_right[self._get_right_diagonal(row, column)],
            self._is_occupied_diagonal_left[self._get_left_diagonal(row, column)],
        )

    def _update_occupancy(
        self, row: int, column: int, is_occupied: OccupancyStatus
    ) -> None:
        self._is_occupied_column[column] = is_occupied
        self._is_occupied_diagonal_right[self._get_right_diagonal(row, column)] = (
            is_occupied
        )
        self._is_occupied_diagonal_left[self._get_left_diagonal(row, column)] = (
            is_occupied
        )

    def _get_right_diagonal(self, row: int, column: int) -> int:
        return row + column

    def _get_left_diagonal(self, row: int, column: int) -> int:
        return (row - column) + self._get_left_diagonal_offset()

    def _get_left_diagonal_offset(self) -> int:
        return self.n - 1

    def _get_number_of_diagonals(self) -> int:
        return self.n * 2 - 1

    def _print_solution(self) -> None:
        print(f'{self.solution_counter}: {" ".join(map(str, self._placed_queen_id))}')
        self.solution_counter += 1


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '-n', type=int, required=True, help='number of queens', dest='N'
    )
    args = parser.parse_args()
    QueenProblemSolver(args.N)

File: rust/queen.rs
use std::env;
use std::sync::atomic::{AtomicU32, Ordering};

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        panic!("Usage: {} <number_of_queens>", args[0])
    }

    let n: i32 = args[1].parse().expect("number_of_queens must be a number");
    solve_queen(n);
}

fn solve_queen(n: i32) {
    if n <= 0 { return }
    let mut occupancy: OccupancyState = OccupancyState::new(n);
    let mut placed_queen_id: Vec<i32> = vec![0; n as usize];

    place_nth_queen(0, n, &mut occupancy, &mut placed_queen_id);
}

struct OccupancyState {
    n: i32,
    is_occupied_column: Vec<bool>,
    is_occupied_right_diagonal: Vec<bool>,
    is_occupied_left_diagonal: Vec<bool>,
}

impl OccupancyState {
    pub fn new(n: i32) -> Self {
        Self {
            n,
            is_occupied_column: vec![false; n as usize],
            is_occupied_right_diagonal: vec![false; Self::get_number_of_diagonals(n) as usize],
            is_occupied_left_diagonal: vec![false; Self::get_number_of_diagonals(n) as usize],
        }
    }

    pub fn update_occupancy(&mut self, row: i32, column: i32, is_occupied: bool) {
        let right_diagonal: i32 = self.get_right_diagonal(row, column);
        let left_diagonal: i32 = self.get_left_diagonal(row, column);

        self.is_occupied_column[column as usize] = is_occupied;
        self.is_occupied_right_diagonal[right_diagonal as usize] = is_occupied;
        self.is_occupied_left_diagonal[left_diagonal as usize] = is_occupied;
    }

    pub fn is_occupied(&self, row: i32, column: i32) -> bool {
        self.is_occupied_column[column as usize]
            || self.is_occupied_right_diagonal[self.get_right_diagonal(row, column) as usize]
            || self.is_occupied_left_diagonal[self.get_left_diagonal(row, column) as usize]
    }

    fn get_number_of_diagonals(n: i32) -> i32 {
        n * 2 - 1
    }

    fn get_right_diagonal(&self, row: i32, column: i32) -> i32 {
        row + column
    }

    fn get_left_diagonal(&self, row: i32, column: i32) -> i32 {
        row - column + self.get_left_diagonal_offset()
    }

    fn get_left_diagonal_offset(&self) -> i32 {
        self.n - 1
    }
}

fn place_nth_queen(
    row: i32,
    n: i32,
    occupancy: &mut OccupancyState,
    placed_queen_id: &mut Vec<i32>,
) {
    for column in 0..n {
        if !occupancy.is_occupied(row, column) {
            placed_queen_id[row as usize] = column;
            occupancy.update_occupancy(row, column, true);

            if row + 1 < n {
                place_nth_queen(row + 1, n, occupancy, placed_queen_id);
            } else {
                print_solution(placed_queen_id);
            }

            occupancy.update_occupancy(row, column, false);
        }
    }
}

static SOLUTION_COUNTER: AtomicU32 = AtomicU32::new(0);

fn print_solution(placed_queen_id: &Vec<i32>) {
    let count = SOLUTION_COUNTER.fetch_add(1, Ordering::SeqCst);
    print!("{}: ", count);
    for column in placed_queen_id.iter() {
        print!("{} ", column);
    }
    println!();
}
